File: php-checks/src/main/java/org/sonar/php/checks/security/LDAPAuthenticatedConnectionCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks.security;

import java.util.Optional;
import java.util.Set;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.CheckUtils;
import org.sonar.php.checks.utils.FunctionUsageCheck;
import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;

@Rule(key = "S4433")
public class LDAPAuthenticatedConnectionCheck extends FunctionUsageCheck {

  private static final String MESSAGE = "Provide username and password to authenticate the connection.";

  @Override
  protected Set<String> lookedUpFunctionNames() {
    return Set.of("ldap_bind");
  }

  @Override
  protected void checkFunctionCall(FunctionCallTree tree) {
    if (argumentIsNullOrEmptyString(tree, "bind_rdn", 1) || argumentIsNullOrEmptyString(tree, "bind_password", 2)) {
      context().newIssue(this, tree, MESSAGE);
    }
  }

  private boolean argumentIsNullOrEmptyString(FunctionCallTree tree, String argumentName, int argumentIndex) {
    Optional<CallArgumentTree> argument = CheckUtils.argument(tree, argumentName, argumentIndex);
    if (argument.isPresent()) {
      ExpressionTree argumentValue = CheckUtils.assignedValue(argument.get().value());
      return CheckUtils.isNullOrEmptyString(argumentValue);
    }
    return true;
  }

}

File: php-checks/src/test/resources/checks/security/LDAPAuthenticatedConnectionCheck.php
<?php

$ldapconn = ldap_connect("foo.com");

$ldapbind = ldap_bind($ldapconn); // Noncompliant {{Provide username and password to authenticate the connection.}}
//          ^^^^^^^^^^^^^^^^^^^^

  LDAP_BIND($ldapconn, "username"); // Noncompliant
//^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  ldap_bind($ldapconn, NULL, NULL); // Noncompliant
//^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

ldap_bind($ldapconn, "username", ""); // Noncompliant

$a = "";
$b = "";
ldap_bind($ldapconn, $a, $b); // Noncompliant

$c = "username";
$d = "";
ldap_bind($ldapconn, $c, $d); // Noncompliant

$e = "user";
$f = "pass";
ldap_bind($ldapconn, $e, $f);


ldap_bind($ldapconn, "username", "password");
ldap_bind($ldapconn, $username, $password); // unknown variables

ldap_bind(link_identifier:$ldapconn, bind_rdn:"username", bind_password:"password");
ldap_bind(link_identifier:"", bind_rdn:"username", bind_password:"password");
ldap_bind(bind_password:"password", bind_rdn:"username");
ldap_bind(bind_password:"", bind_rdn:"username", link_identifier:$ldapconn); // Noncompliant
ldap_bind(link_identifier:$ldapconn, bind_rdn:"username"); // Noncompliant
ldap_bind(link_identifier:$ldapconn, bind_password:"password"); // Noncompliant

function foobar($cond) {
  $g = "";
  $h = "";
  if ($cond) {
    $g = "user";
    $h = "pass";
  }
  ldap_bind($ldapconn, $g, $h); // FN
}
